pkg/server/api/agent/v1/service.go

```diff
@@ -47,7 +56,47 @@ func (s *Service) DeleteAgent(ctx context.Context, req *agent.DeleteAgentRequest
 }
 
 func (s *Service) BanAgent(ctx context.Context, req *agent.BanAgentRequest) (*empty.Empty, error) {
-	return nil, status.Error(codes.Unimplemented, "method not implemented")
+	log := rpccontext.Logger(ctx)
+
+	id, err := api.IDFromProto(req.Id)
+	if err != nil {
+		log.WithError(err).Error("Invalid request: invalid SPIFFE ID")
+		return nil, status.Errorf(codes.InvalidArgument, "invalid SPIFFE ID: %v", err)
+	}
+
+	log = log.WithField(telemetry.SPIFFEID, id.String())
+
+	if !idutil.IsAgentPath(id.Path()) {
+		log.Error("Invalid request: not an agent ID")
+		return nil, status.Error(codes.InvalidArgument, "not an agent ID")
+	}
+
+	if !id.MemberOf(s.td) {
+		log.Error("Invalid request: cannot ban an agent that does not belong to this trust domain")
+		return nil, status.Errorf(codes.InvalidArgument, "cannot ban an agent that does not belong to this trust domain")
+	}
+
+	// The agent "Banned" state is pointed out by setting its
+	// serial numbers (current and new) to empty strings.
+	_, err = s.ds.UpdateAttestedNode(ctx, &datastore.UpdateAttestedNodeRequest{
+		SpiffeId: id.String(),
+		InputMask: &common.AttestedNodeMask{
+			CertSerialNumber:    true,
+			NewCertSerialNumber: true,
+		},
+	})
+
+	switch status.Code(err) {
+	case codes.OK:
+		log.Debug("Agent banned")
```

Comment:
How would you feel about making this `INFO` level? I think this operation is important to audit via normal logging.

Reply:
Sounds good!